### DISCUSSION TITLE
Improve Home Assistant integration

### DISCUSSION THREAD

**Discussion by StarkJohan** [UtilitechAS/amsreader-firmware#477] [Q&A]
I'm looking to once and for all get the Home Assistant energy dashboard to work as intended. I'm not sure about how to solve the problem so I thought a discussion/q&a might be a good way to go about it.

My kamstrup meter reports the total consumption once an hour, just after the hour as most meters seems to do. This brings a few issues.

1. Once, every few days, HA does not pick up the MQTT message from AMSReader and thus no hourly data is reported for this hour. Everything consumed during this hour is then added to the next hour in HA resulting in bad reporting and price calculating errors (using hourly rates).
2. As the report is recieved by HA slightly after the full hour the consumption is always one hour delayed in the HA energy dashboard. I know this is a known issue.

As far as I can tell these issues can only be resolved by more frequent reporting of the "Accumulated active import". I might be wrong on this but I can't really see any better solution.

Would it be possible to get more frequent reporting from AMSReader? I know that such "in between" reports would be calculated values but they should be more than accurate enough to improve the current HA situation vastly. 
At this point everything is one hour off and in some cases two hours off (when the hourly report drops out).

Looking forward to your input on this.

**Comment by ArnieO** [comment]
The "whole hour" payload from Kamstrup comes in between two "normal" 10-second-interval payloads, it does not replace one of them. 

If I understand your proposal correctly, you want to send accumulated energy estimated by the amsreader a few seconds before whole hour to "compensate" for potentially missing message (with "wrong" timestamp) that arrives shortly after whole hour. 

Should be technically feasible - but could there be a drawback doing that? I'm not using HA, so those of you using HA need to consider if there might be a pitfall here.

**Comment by StarkJohan** [comment]
I'm thinking that sending this "calculated accumulated sum" every 5 or 10 minutes would work better, not only just before the full hour as this also might be lost. Should one measurement be missed with this higher frequency for some reason (HA reboot, errors, whatever) the data would only be 5 or 10 minutes off. This would bo totally acceptable, especially considering we're 1h off today.

A problem that could arise is that the calculated sum is slightly higher at the end of the hour than the actual reported value recieved from the meter a few minutes later. As far as I know this would throw a log warning in HA but no worse than that. 

Can't see any real issues on the HA side. I haven't studied the AMSReader sources but calculating total energy from the "live" 10s reports should be fairly accurate and probably not to tricky to implement.

**Reply by StarkJohan** [reply]
Here's an example of my HA and AMSReader data today. Note the cars charging between 02-04 reported as 03-05 in HA and two missing hours in the evening.

![Screenshot from 2023-03-10 21-36-25](https://user-images.githubusercontent.com/8789624/224423629-60582e05-08dc-4c67-b3ba-8e7ee6efbf6b.png)
![Screenshot from 2023-03-10 21-35-58](https://user-images.githubusercontent.com/8789624/224423634-1829876f-f0de-4f0a-bd69-b18a737bf42b.png)

**Comment by gskjold** [comment]
This firmware will only send "real" data when it arrives, except for the realtime data which is calculated. There could be a extra sensor sending the calculated total kwh reading every time it is updated. This sensor may not be reliable though, so I'm not sure it is a good idea

**Reply by StarkJohan** [reply]
Do you have any other ideas on how to get a more reliable relationship between amdreader and home assistant? 

I would prefer having more frequent updates in addition to the real updates even if the calculated values may be a bit off at the end on the hour. The calculated value will still be synced to the real value every hour.

How big of an error/reliability issue do you think we might experience?

**Reply by gskjold** [reply]
I'm not that familiar of the inner workings of HA, so this is the best solution I can come up with. The deviation varies a bit between the meters, for mine it is less than 100wh, but I've heard some experience more

**Reply by StarkJohan** [reply]
Would adding such a sensor to amsreader be a big task or is it doable in the forseable future? I understand that it might not at all be in line with your plans.

I guess i could create a similar calculated value in HA but that would be a less dynamic and "nice" solution.

**Reply by gskjold** [reply]
Shouldn't be too much work. There are several requests and issues for HA integration, so I will put this one on the list as well
